src/components/implementation/tests/unit_fprr/unit_fprr.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2016, Phani Gadepalli and Gabriel Parmer, GWU, gparmer@gwu.edu.
+ *
+ * This uses a two clause BSD License.
+ */
+
+#include <cos_defkernel_api.h>
+#include <llprint.h>
+#include <res_spec.h>
+#include <sl.h>
+
+/* Ensure this is the same as what is in sl_mod_fprr.c */
+#define SL_FPRR_NPRIOS 32
+
+#define LOWEST_PRIORITY (SL_FPRR_NPRIOS - 1)
+
+#define LOW_PRIORITY (LOWEST_PRIORITY - 1)
+#define HIGH_PRIORITY (LOWEST_PRIORITY - 10)
+
+int lowest_was_scheduled = 0;
+
+void
+low_thread_fn()
+{
+	lowest_was_scheduled = 1;
+	sl_thd_free(sl_thd_curr());
+}
+
+void
+high_thread_fn()
+{
+	struct sl_thd *low_thread = sl_thd_alloc(low_thread_fn, NULL);
+	sl_thd_param_set(low_thread, sched_param_pack(SCHEDP_PRIO, LOW_PRIORITY));
+
+	cycles_t deadline = sl_now() + sl_usec2cyc(10 * 1000 * 1000);
+	while (sl_now() < deadline) {}
+	assert(!lowest_was_scheduled);
+	sl_thd_free(sl_thd_curr());
+}
+
+void
+test_highest_is_scheduled(void)
+{
+	struct sl_thd *high_thread = sl_thd_alloc(high_thread_fn, NULL);
+	sl_thd_param_set(high_thread, sched_param_pack(SCHEDP_PRIO, HIGH_PRIORITY));
+
+	cycles_t wakeup = sl_now() + sl_usec2cyc(1000 * 1000);
+	sl_thd_block_timeout(0, wakeup);
+}
+
+int thd1_ran = 0;
+int thd2_ran = 0;
+
+void
+thd1_fn()
+{
+	thd1_ran = 1;
+	while(1);
+}
+
+void
+thd2_fn()
+{
+	thd2_ran = 1;
+	while(1);
```

Comment:
Is it possible to have more useful output here? 
Like, lets say, time between switches, that could help in identifying if the timeouts are working correctly, of course on baremetal! :)